staging/src/kubevirt.io/api/core/v1/types.go

```diff
@@ -2420,6 +2423,20 @@ type KubeVirtConfiguration struct {
 	AutoCPULimitNamespaceLabelSelector *metav1.LabelSelector `json:"autoCPULimitNamespaceLabelSelector,omitempty"`
 	// LiveUpdateConfiguration holds defaults for live update features
 	LiveUpdateConfiguration *LiveUpdateConfiguration `json:"liveUpdateConfiguration,omitempty"`
+
+	// VMRolloutStrategy defines how changes to a VM object propagate to its VMI
+	VMRolloutStrategy *VMRolloutStrategy `json:"vmRolloutStrategy,omitempty"`
+}
+
+type VMRolloutStrategy struct {
+	LiveUpdate *RolloutStrategyLiveUpdate `json:"liveUpdate,omitempty"`
+	Stage      *RolloutStrategyStage      `json:"stage,omitempty"`
```

Comment:
This follows what's done elsewhere, and unifying the structs would be a big change that wouldn't get us much, I'd rather keep it like that if it's ok.

Reply:
I am ok with it

Reply:
Why do we need any struct at all? During the discussion of https://github.com/kubevirt/community/blob/main/design-proposals/vm-rollout-strategy/vm-rollout-strategy.md#kubevirt-cr I expected to see here plain `const` strings, such as the ones we have for `EvictionStrategy`.

Reply:
Please see https://github.com/kubevirt/kubevirt/pull/11054#discussion_r1466501632
I believe @xpivarc and @acardace were part of the pro-struct advocates, hopefully they can add some info here.
I don't have a strong opinion either way, please just know that changing from one to the other takes about a whole day of work...

Reply:
Can anyone describe an imaginable idea about how a strategy can/should be extended?

Committing on an API is hard, but choosing extensibility without a clear need is costly. It's more complex for users that we can image, and it opens the door for mistakes (such as supplying no strategy, or both strategies) and then to validations, and then to tests for validation, which are not needed if the type suits its planned usage.

@jean-edouard please keep the `struct` if you are convinced that it adds value. But if we don't recall the value of an API knob or complexity, we should opt for the simpler solution.

Reply:
> Can anyone describe an imaginable idea about how a strategy can/should be extended?

I think the idea is that we don't know what we'll need to add in the future...
But for example, if/when network/storage join the party, maybe we'll want to make that optional at first to give users time to get used to the new behavior.
More generally, maybe some users will want all fields to be live-updatable except a few specific ones.

> 
> Committing on an API is hard, but choosing extensibility without a clear need is costly. It's more complex for users that we can image, and it opens the door for mistakes (such as supplying no strategy, or both strategies) and then to validations, and then to tests for validation, which are not needed if the type suits its planned usage.

I (obviously) agree that API should be kept as simple a possible. However here, string vs struct doesn't really have an impact on complexity either way.
In fact, using structs allows for things like auto-complete and rejection of invalid entries without needing a webhook, so I'd say it's slightly less complex!

> 
> @jean-edouard please keep the `struct` if you are convinced that it adds value. But if we don't recall the value of an API knob or complexity, we should opt for the simpler solution.

There have been so many discussions on so many channels about this PR that I unfortunately don't remember who said what. However, I remember a bunch of engineers unanimously argued that "structs are always better". @vladikr do you recall such a conversation, or have an opinion here?
I personally don't feel strongly about either approach, I just don't want to waste a day reverting to strings for nothing (or worse, do it then have the pro-structs wake up and request a switch back!)

Reply:
@stu-gott also added offline (this time I'm keeping a trace!) that using structs allows us to individually document each option.
That reminded me that I forgot to do so. Fixed now.

Reply:
imho the documentation is not a strong enough motivation for the added complexity. It is better to have a tight API that does not need close reading of its documentation, over adding `VMRolloutStrategy struct` with two fields, and having to document that you can set only one of them.

Reply:
So. My biggest issue with this: It deviates fom the design discussion. And I did all my thinking during the design discussion, which was converged  - not with structs, but with consts.
I'd need to redo the whole discussion here in order to recall my thoughts.

In future: Let's please stick to what we discussed and agreed on - it is otherwse at least making my life more difficult.

For this case: I know that we have the "struct approach" in a couple of places.
Can we move back to consts, unless we have an example of why a struct is needed?

Reply:
FWIW, structs are superior in such cases because we don't know how the future layout will look like and what else will we need here. However, changing this api later may be tricky. 
We are limiting ourselves by not adopting an empty struct.
For reference see `DaemonSetUpdateStrategy` in https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec
> DaemonSetUpdateStrategy is a struct used to control the update strategy for a DaemonSet.